File: matlab/Figure7_1.m
% Author: Kenji Kashima
% Date  : 2025/09/11

rng(24);  

n_x = 100;                           % n_x - number of x-grid for plot
x_p = linspace(0,1,n_x);             % x grid points for plot

% Feature mapping size
n_f = numel(phi(0));                 % number of features

% matrix phi for plot
Phi_p = zeros(n_f, n_x);
for i = 1:n_x
    Phi_p(:, i) = phi(x_p(i));
end

% Draw figures
figure7_1a(20, x_p, Phi_p, n_f);
sigma_sq = [0.5, 100, 10^(-6) ];     % standard deviation sigma = 0.5, 10, 10^{-3}
figure7_1b(sigma_sq, 20, 20, x_p, Phi_p, n_f);


% ======================= Functions =======================

function v = phi(x)
    % Feature mapping
    % x - input
    v = [1; x; x.^2; x.^3; x.^4; x.^5; x.^6; x.^7; x.^8; x.^9];
end

function y = f_true(x)
    % x - input
    y = 2*sin(5*x);
end

function figure7_1a(n_sample, x_p, Phi_p, n_f)
    % n_sample - number of sample functions

    figure();
    plot(x_p, zeros(size(x_p)), 'LineWidth', 3); hold on;

    for k = 1:n_sample
        para = randn(n_f, 1);
        fx = (para.' * Phi_p);                  % 1×n_x
        plot(x_p, fx, 'LineWidth', 0.5, 'Color', [0.7,0.7,0.7]);
    end

    xlabel('$\rm x$','Interpreter','latex');
    ylabel('$f({\rm x})$','Interpreter','latex');
    xlim([0 1]); grid on;
    set(gca,'FontSize',12);
end

function figure7_1b(sigma_sq, n_sample, s_bar, x_p, Phi_p, n_f)
    % sigma_sq - list of squared SD
    % n_sample - number of sample functions
    % s_bar - number of data

    x = linspace(0,1,s_bar);                          % x for data
    Phi = zeros(n_f, s_bar);                           % matrix phi in (7.10)
    for s = 1:s_bar
        Phi(:, s) = phi(x(s));
    end
    y = arrayfun(@(xs) f_true(xs), x)' + randn(s_bar,1); % noisy observation (column)

    size_sigma_sq = numel(sigma_sq);
    mean = zeros(n_f, size_sigma_sq);
    cov  = zeros(n_f, size_sigma_sq*n_f);

    % Compute mean and cov blocks
    for l = 1:size_sigma_sq
        % tmp = I - Phi * inv(Phi'Phi + sigma^2 I) * Phi'
        % (keep variable names and structure)
        A = (Phi.' * Phi) + sigma_sq(l) * eye(s_bar);
        % use matrix division instead of inv for stability
        tmp = eye(n_f) - Phi / A * Phi.';              % n_f × n_f
        cov(:, (n_f*(l-1)+1) : n_f*l) = 0.5 * (tmp + tmp.');

        % mean = Phi * inv(Phi'Phi + sigma^2 I) * y
        mean(:, l) = Phi / A * y;                      % n_f × 1
    end

    % Plot
    figure(); hold on;

    yhat1 = (mean(:,1).') * Phi_p;  % 1×n_x
    plot(x_p, yhat1, 'LineWidth', 2, 'DisplayName', '$\sigma=0.5$', 'Color',[0 0.4470 0.7410]);

    if size_sigma_sq >= 2
        yhat2 = (mean(:,2).') * Phi_p;
        plot(x_p, yhat2, 'LineWidth', 2, 'DisplayName', '$\sigma=10$');
    end
    if size_sigma_sq >= 3
        yhat3 = (mean(:,3).') * Phi_p;
        plot(x_p, yhat3, 'LineWidth', 2, 'DisplayName', '$\sigma=10^{-3}$');
    end

    % Sampling from the posterior (first block)
    Cblk = cov(:, 1:n_f);                      % n_f × n_f
    [R, ~] = chol(Cblk, 'lower');

    for k = 1:n_sample
        z = randn(n_f,1);
        para = mean(:,1) + R * z;              % multivariate normal
        fx = (para.' * Phi_p);
        plot(x_p, fx, 'LineWidth', 0.3, 'Color', [0.7,0.7,0.7], 'HandleVisibility','off');
    end

    scatter(x, y, 36, 'filled', 'DisplayName', '${\rm y}_s$');
    legend('Interpreter','latex');

    xlabel('$\rm x$','Interpreter','latex');
    ylabel('$f({\rm x})$','Interpreter','latex');
    xlim([0 1]); grid on;
    set(gca,'FontSize',12);
end

File: python/Figure7_1.py
# Author: Kenji Kashima
# Date  : 2025/04/01

import numpy as np
import matplotlib.pyplot as plt
import sys
sys.path.append("./")
import config

np.random.seed(24)

def phi(x:float)->np.ndarray:
    '''
        Feature mapping
        x - input
    '''
    return  np.array([1, x, x**2, x**3, x**4, 
                      x**5, x**6, x**7, x**8, x**9 ])

n_f = phi(0).size



def f_true(x:float)->float:
    '''
        x - input
    '''
    return 2*np.sin(5*x)

def figure7_1a(n_x:int = 100, n_sample:int=20):
    '''
        n_x - number of x-grid for plot
        n_sample - number of sample functions
    '''
    figsize = config.global_config(type=1)

    x_p = np.linspace(0,1,n_x)
    X = np.zeros([n_f,n_x]) 
    for i in range(n_x):
        X[:,i] = phi(x_p[i]) 

    plt.figure(figsize=figsize)
    plt.plot(x_p, np.zeros_like(x_p), linewidth=3)

    for _ in range (n_sample):
        theta = np.random.randn(n_f)      
        fx =  theta @ X                 
        plt.plot(x_p,fx,linewidth= 0.5, color=[0.7,0.7,0.7])    
    
    plt.xlabel(r'$\rm x$')
    plt.ylabel(r'$f({\rm x})$')
    plt.xlim([0,1])
    plt.grid()
    plt.tight_layout()
    plt.savefig("./figures/Figure7_1a.pdf")
    plt.show()

# learning from n_data in one sample trajectory with hyperparameter sigma_sq
def learn_theta(sigma_sq,n_data):
    '''
        sigma_sq - simulation for several weights (=standard deviation)^2
        n_data - number of data
    '''
    x = np.linspace(0,1,n_data)
    X = np.zeros([n_f,n_data])
    y=np.zeros(n_data)
    for s in range(n_data):
        X[:,s] = phi(x[s])
        y[s] = f_true(x[s]) + np.random.randn()

    size_sigma_sq = len(sigma_sq)
    mean = np.zeros([n_f,size_sigma_sq])
    cov = np.zeros([n_f,size_sigma_sq*n_f])

    for l in range(size_sigma_sq):
        tmp = np.eye(n_f) - X @ np.linalg.inv(X.T @ X+sigma_sq[l] * np.eye(n_data)) @ X.T
        cov[:,n_f*l:n_f*(l+1)] = (tmp+tmp.T)/2
        mean[:,l]  = X @ np.linalg.inv(X.T @ X + sigma_sq[l] * np.eye(n_data)) @ y.T
    return mean, cov, x, y

def figure7_1b(n_x:int = 100, n_sample:int=20, n_data:int=8):
    '''
        n_x - numer of x-grid for plot
        n_sample - number of sample functions
        n_data - number of data
    '''
    figsize = config.global_config(type=1)
    
    x_p = np.linspace(0,1,n_x)
    X = np.zeros([n_f,n_x]) 
    for i in range(n_x):
        X[:,i] = phi(x_p[i]) 
    sigma_sq = [0.25,100,10**(-6)]  # standard deviation sigma = 0.5, 10^3, 10^{-3}
    mu,Sigma,x,y = learn_theta(sigma_sq,n_data)

    plt.figure(figsize=figsize)

    plt.plot(x_p, (mu[:,0] @ X).flatten(), linewidth= 2,label=r'$\sigma=0.5$',zorder=10)
    plt.plot(x_p, (mu[:,1] @ X).flatten(), linewidth= 2,label=r'$\sigma=10$')
    plt.plot(x_p, (mu[:,2] @ X).flatten(), linewidth= 2,label=r'$\sigma=10^{-3}$')

    for _ in range (n_sample):
        theta = np.random.multivariate_normal(mu[:,0],Sigma[:,0:n_f])      
        fx =  theta @ X                   
        plt.plot(x_p,fx,linewidth= 0.3,color=[0.7,0.7,0.7])

    plt.scatter(x,y,marker='o',label=r'${\rm y}_s$')
    plt.legend()
    
    plt.xlabel(r'$\rm x$')
    plt.ylabel(r'$f({\rm x})$')
    plt.xlim([0,1])
    plt.grid()
    plt.tight_layout()
    plt.savefig("./figures/Figure7_1b.pdf")
    plt.show()

if __name__ == '__main__':
    figure7_1a(n_x=100,n_sample=20)
    figure7_1b(n_x=100,n_sample=20,n_data=20)
